### PR TITLE
New logging implementation.

### PR DESCRIPTION
**Pull request by carlhampuswall**
New implementation for logging easier to log with different levels and allows for adding origin of log message for future debugging/troubleshooting purposes etc.

### CODE REVIEW

File: firmware/src/configuration.cpp [hunk 1]
```diff
@@ -107,42 +107,41 @@ bool Configuration::saveToDisk()
...
-        char buf[200];
-        snprintf(buf, sizeof(buf), "Encoding failed: %s", PB_GET_ERROR(&stream));
-        log(buf);
+        char buf_[200];
+        snprintf(buf_, sizeof(buf_), "Encoding failed: %s", PB_GET_ERROR(&stream));
+        LOGE(buf_);
         return false;
     }
-    ESP_LOGD("1", "save to disk");
+    LOGD("1 Save to disk");
```
Comment (brushknight) [inline firmware/src/configuration.cpp:115]: can you remove those 1-5 save to disk, they were for debugging and we can drop them at this point

File: firmware/src/logging.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        LOG(log_level, true, ##__VA_ARGS__); \
+    } while (0)
+
+class Logging
+{
+
+public:
+    static Logging &getInstance()
+    {
+        static Logging instance;
```
Comment (brushknight) [inline firmware/src/logging.h:54]: shouldn't you have a check that instance is already created?

File: firmware/src/serial/proto_helpers.h [hunk 3]
```diff
@@ -4,25 +4,12 @@
...
-        && first.min_position == second.min_position
-        && first.max_position == second.max_position
-        && first.position_width_radians == second.position_width_radians
-        && first.snap_point == second.snap_point
-        && first.sub_position_unit == second.sub_position_unit
-        && strcmp(first.text, second.text) == 0
-        && first.detent_positions_count == second.detent_positions_count
-        && memcmp(first.detent_positions, second.detent_positions, first.detent_positions_count * sizeof(first.detent_positions[0]))
-        && first.snap_point_bias == second.snap_point_bias;
+bool config_eq(PB_SmartKnobConfig &first, PB_SmartKnobConfig &second)
```
Comment (brushknight) [inline firmware/src/serial/proto_helpers.h:7]: I like the multi-line approach more, is there a reason why we removed it?

**Review by brushknight — COMMENTED** [review]
Looks good, added a few comments

File: firmware/src/logging.h (hunk 2 above)
Comment (carlhampuswall) [reply firmware/src/logging.h:54]: since `static Logging instance;` is static we dont have to

File: firmware/src/serial/proto_helpers.h (hunk 3 above)
Comment (carlhampuswall) [reply firmware/src/serial/proto_helpers.h:7]: My formatting rules... do u know what i need to change? couldn't find it with a quick google search :/

**Review by brushknight — APPROVED** [review]

**Pull request merged**